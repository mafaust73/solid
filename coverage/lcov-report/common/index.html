<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for common</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../prettify.css" />
    <link rel="stylesheet" href="../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../index.html">All files</a> common
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">95.05% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>192/202</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">80.26% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>61/76</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">94.34% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>50/53</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">95.36% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>185/194</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="bspNode.js"><a href="bspNode.js.html">bspNode.js</a></td>
	<td data-value="86.49" class="pic high"><div class="chart"><div class="cover-fill" style="width: 86%;"></div><div class="cover-empty" style="width:14%;"></div></div></td>
	<td data-value="86.49" class="pct high">86.49%</td>
	<td data-value="74" class="abs high">64/74</td>
	<td data-value="73.91" class="pct medium">73.91%</td>
	<td data-value="46" class="abs medium">34/46</td>
	<td data-value="78.57" class="pct medium">78.57%</td>
	<td data-value="14" class="abs medium">11/14</td>
	<td data-value="87.14" class="pct high">87.14%</td>
	<td data-value="70" class="abs high">61/70</td>
	</tr>

<tr>
	<td class="file empty" data-value="common.js"><a href="common.js.html">common.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="csg.js"><a href="csg.js.html">csg.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="37" class="abs high">37/37</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="35" class="abs high">35/35</td>
	</tr>

<tr>
	<td class="file high" data-value="plane.js"><a href="plane.js.html">plane.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="48" class="abs high">48/48</td>
	<td data-value="86.36" class="pct high">86.36%</td>
	<td data-value="22" class="abs high">19/22</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="48" class="abs high">48/48</td>
	</tr>

<tr>
	<td class="file high" data-value="polygon.js"><a href="polygon.js.html">polygon.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="9" class="abs high">9/9</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="7" class="abs high">7/7</td>
	</tr>

<tr>
	<td class="file high" data-value="vector.js"><a href="vector.js.html">vector.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="29" class="abs high">29/29</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="15" class="abs high">15/15</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="29" class="abs high">29/29</td>
	</tr>

<tr>
	<td class="file high" data-value="vertex.js"><a href="vertex.js.html">vertex.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Jul 01 2018 12:42:36 GMT+0200 (CEST)
</div>
</div>
<script src="../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../sorter.js"></script>
<script src="../block-navigation.js"></script>
</body>
</html>
