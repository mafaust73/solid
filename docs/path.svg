<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg
   xmlns:dc="http://purl.org/dc/elements/1.1/"
   xmlns:cc="http://creativecommons.org/ns#"
   xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
   xmlns:svg="http://www.w3.org/2000/svg"
   xmlns="http://www.w3.org/2000/svg"
   version="1.1"
   id="svg2"
   viewBox="0 0 312.53784 312.53784"
   height="88.205124mm"
   width="88.205124mm">
  <defs
     id="defs4">
    <marker
       orient="auto"
       refY="0"
       refX="0"
       id="marker4462"
       style="overflow:visible">
      <path
         id="path4464"
         style="fill:#000000;fill-opacity:1;fill-rule:evenodd;stroke:#000000;stroke-width:0.625;stroke-linejoin:round;stroke-opacity:1"
         d="M 8.7185878,4.0337352 -2.2072895,0.01601326 8.7185884,-4.0017078 c -1.7454984,2.3720609 -1.7354408,5.6174519 -6e-7,8.035443 z"
         transform="matrix(-1.1,0,0,-1.1,-1.1,0)" />
    </marker>
    <marker
       style="overflow:visible"
       id="Arrow2Lend"
       refX="0"
       refY="0"
       orient="auto">
      <path
         transform="matrix(-1.1,0,0,-1.1,-1.1,0)"
         d="M 8.7185878,4.0337352 -2.2072895,0.01601326 8.7185884,-4.0017078 c -1.7454984,2.3720609 -1.7354408,5.6174519 -6e-7,8.035443 z"
         style="fill:#000000;fill-opacity:1;fill-rule:evenodd;stroke:#000000;stroke-width:0.625;stroke-linejoin:round;stroke-opacity:1"
         id="path4165" />
    </marker>
  </defs>
  <metadata
     id="metadata7">
    <rdf:RDF>
      <cc:Work
         rdf:about="">
        <dc:format>image/svg+xml</dc:format>
        <dc:type
           rdf:resource="http://purl.org/dc/dcmitype/StillImage" />
        <dc:title></dc:title>
      </cc:Work>
    </rdf:RDF>
  </metadata>
  <g
     transform="translate(-224.56693,-96.929126)"
     id="layer1">
    <path
       id="path4138"
       d="m 260,252.3622 240,0"
       style="fill:none;fill-rule:evenodd;stroke:#000000;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1;marker-end:url(#Arrow2Lend)" />
    <path
       style="fill:none;fill-rule:evenodd;stroke:#000000;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1;marker-end:url(#marker4462)"
       d="m 380,374.0339 0,-240"
       id="path4420" />
    <text
       id="text4494"
       y="271.3779"
       x="491.73346"
       style="font-style:normal;font-weight:normal;font-size:12.5px;line-height:125%;font-family:sans-serif;letter-spacing:0px;word-spacing:0px;fill:#000000;fill-opacity:1;stroke:none;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1"
       xml:space="preserve"><tspan
         y="271.3779"
         x="491.73346"
         id="tspan4496">U</tspan></text>
    <text
       id="text4498"
       y="142.45804"
       x="360.63589"
       style="font-style:normal;font-weight:normal;font-size:12.5px;line-height:125%;font-family:sans-serif;letter-spacing:0px;word-spacing:0px;fill:#000000;fill-opacity:1;stroke:none;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1"
       xml:space="preserve"><tspan
         y="142.45804"
         x="360.63589"
         id="tspan4500">V</tspan></text>
    <path
       id="path4532"
       d="m 380,252.3622 25,0 0,-5 -20,-5 0,-35 20,-5 10,-15 5,-20 -4.64528,0.16734 -5.66934,19.83266 -7.34269,11.50602 L 380,204.22294"
       style="fill:none;fill-rule:evenodd;stroke:#ff01ce;stroke-width:1px;stroke-linecap:butt;stroke-linejoin:miter;stroke-opacity:1" />
  </g>
</svg>
