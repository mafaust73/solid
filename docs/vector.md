# Vector

Create a vector using `new Vector(x, y, z)`, e.g. `new Vector(1, 0, 0)`;

## Operations

All operations are immutable, i.e. returning always a new vector.

| Operation                        | Method                               |
| -------------------------------- | ------------------------------------ |
| a + b                            | a.plus(b)                            |
| a - b                            | a.minus(b)                           |
| a \* 7                           | a.times(7)                           |
| a / 7                            | a.dividedBy(7)                       |
| a ⋅ b                            | a.dot(b)                             |
| a ⨯ b                            | a.cross(b)                           |
| -a                               | a.negated()                          |
| Interpolate between vectors      | a.lerp(b, t) with 0 ≤ t ≤ 1          |
| Length of a vector               | a.length()                           |
| Normalize a vector to length = 1 | a.unit()                             |
| Rotate around y-axis             | a.rotateY(alpha) with 0 ≤ alpha ≤ 2π |
| Return a digital copy            | a.clone()                            |

## Constants

The library provides constants for

| Constant          | Value                                    |
| ----------------- | ---------------------------------------- |
| ZERO              | new Vector(0, 0, 0)                      |
| ONE               | new Vector(1, 1, 1)                      |
| UNIT_X            | new Vector(1, 0, 0)                      |
| UNIT_Y            | new Vector(0, 1, 0)                      |
| UNIT_Z            | new Vector(0, 0, 1)                      |
| POSITIVE_INFINITY | new Vector(Infinity, Infinity, Infinity) |
| NEGATIVE_INFINITY | new Vector(Infinity, Infinity, Infinity) |
