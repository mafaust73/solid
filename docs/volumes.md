# Volumes

## Simple

### Cube

A cube is described by:

- center : Vector
- size : Vector.

```javascript
import { Cube, ZERO, ONE } from "solid";

let center = ZERO;
let size = ONE;

let cube = new Cube(center, size);
```

![Cube](cube.png)

### Sphere

A sphere is described by:

- center : Vector
- radius : Number
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Sphere, ZERO, ONE } from "solid";

let center = ZERO;
let radius = 1;
let resolution = 10;

let sphere = new Sphere(center, radius, resolution);
```

![Sphere](sphere.png)

### Cylinder

A cylinder is described by:

- start : Vector
- end : Vector
- radius : Number
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Cylinder, ZERO, ONE, UNIT_Y } from "solid";

let start = ZERO;
let end = UNIT_Y;
let radius = 1;
let resolution = 10;

let cylinder = new Cylinder(start, end, radius, resolution);
```

![Cylinder](cylinder.png)

### Cone

A cone is described by:

- start : Vector
- end : Vector
- radiusStart : Number
- radiusEnd : Number
- closed : Boolean
- resolution : Integer  
  With `resolution` you can control the granularity of the approximation.

```javascript
import { Cone, ZERO, ONE, UNIT_Y } from "solid";

let start = ZERO;
let end = UNIT_Y;
let radiusStart = 1;
let radiusEnd = 0.2;
let resolution = 10;

let cone = new Cone(start, end, radiusStart, radiusEnd, resolution);
```

![Cone](cone.png)

## Complex

### Lathe

Creates a rotational object:

- path : Path the path in the xy plane
- resolution : Integer the path resolution
- slices Integer the number of slices around the rotational axis

```javascript
import { Lathe, Path } from "solid";

let path = new Path()...;
let slices = 32;

let lathe = new Lathe(path, resolution, slices);
```

![Lathe](lathe.png)

### Twist

Creates a rotational object:

- path : Path the path in the xz plane
- aspect : Number the amount of twist per height unit
- resolution : Integer the path resolution
- stacks Integer the number of stacks along the y-axis

![Twist](twist.png)

### Path

Paths are building the base for many volumes. Paths are build in an abstract U-V-Space and translated according to the given axis.

The following image shows a simple path for creating a glas object.
![path](path.svg)

A path is created using a flow builder pattern:

```javascript
let path = new Path()
  .moveTo(0, 0)
  .lineTo(10, 0)
  .lineTo(10, 2);
```

| Operation    | Description                       |
| ------------ | --------------------------------- |
| moveTo(u, v) | Must be the first item of a path. |
| lineTo(u, v) | Create a linear segment           |

TODO: arc and bezier paths

## Combine geometries

Combine two geometries by union, subtract and intersect.

```javascript
import { Cube, Sphere, ZERO, ONE } from "solid";
let cube = new Cube(ZERO, ONE);
let sphere = new Sphere(ZERO, 1.2, 32);

let union = cube.union(sphere);
let subtract = cube.subtract(sphere);
let intersect = cube.intersect(sphere);
```

| union               | subtract                  | intersect                   |
| ------------------- | ------------------------- | --------------------------- |
| ![union](union.png) | ![subtract](subtract.png) | ![intersect](intersect.png) |
