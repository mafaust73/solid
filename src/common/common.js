export * from "./csg";
export * from "./plane";
export * from "./polygon";
export * from "./vector";
export * from "./vertex";
