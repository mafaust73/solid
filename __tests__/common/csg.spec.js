import { Vector, UNIT_X, UNIT_Y } from "../../src/common/vector";
import { Cube } from "../../src/volumes/cube";
import { Cylinder } from "../../src/volumes/cylinder";
import { Cone } from "../../src/volumes/cone";
import { Wavefront } from "../../src/io/wavefront";
import { Path } from "../../src/misc/path";
import { Twist } from "../../src/volumes/twist";
import { Lathe } from "../../src/volumes/lathe";

import fs from "fs";

describe("csg", () => {
  test("invert, union, subtract, intersect", () => {
    const a = new Cube(new Vector(-0.5, -0.5, -0.5), new Vector(1, 1, 1));
    const b = new Cube(new Vector(0.5, 0.5, 0.5), new Vector(1, 1, 1));

    const alias = new Wavefront();
    expect(alias.write(a.union(b))).toMatchSnapshot();
    expect(alias.write(a.subtract(b))).toMatchSnapshot();
    expect(alias.write(a.intersect(b))).toMatchSnapshot();
    expect(alias.write(a.invert())).toMatchSnapshot();
  });
});
