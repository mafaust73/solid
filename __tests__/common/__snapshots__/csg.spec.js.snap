// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csg invert, union, subtract, intersect 1`] = `
"v -1.5 -1.5 -1.5
v -1.5 -1.5 0.5
v -1.5 0.5 0.5
v -1.5 0.5 -1.5
v 0.5 -1.5 -1.5
v 0.5 -0.5 -1.5
v 0.5 -0.5 0.5
v 0.5 -1.5 0.5
v 0.5 -0.5 -1.5
v 0.5 0.5 -1.5
v 0.5 0.5 -0.5
v 0.5 -0.5 -0.5
v 1.5 -0.5 -0.5
v 1.5 1.5 -0.5
v 1.5 1.5 1.5
v 1.5 -0.5 1.5
v 0.5 -0.5 -0.5
v 1.5 -0.5 -0.5
v 1.5 -0.5 1.5
v 0.5 -0.5 1.5
v 0.5 1.5 1.5
v 1.5 1.5 1.5
v 1.5 1.5 -0.5
v 0.5 1.5 -0.5
v 0.5 1.5 -0.5
v 1.5 1.5 -0.5
v 1.5 -0.5 -0.5
v 0.5 -0.5 -0.5
v 0.5 -0.5 1.5
v 1.5 -0.5 1.5
v 1.5 1.5 1.5
v 0.5 1.5 1.5
v -1.5 -1.5 -1.5
v -0.5 -1.5 -1.5
v -0.5 -1.5 0.5
v -1.5 -1.5 0.5
v -0.5 -1.5 -1.5
v 0.5 -1.5 -1.5
v 0.5 -1.5 0.5
v -0.5 -1.5 0.5
v -1.5 0.5 -1.5
v -1.5 0.5 0.5
v -0.5 0.5 0.5
v -0.5 0.5 -1.5
v 0.5 0.5 -0.5
v 0.5 0.5 -1.5
v -0.5 0.5 -1.5
v -0.5 0.5 -0.5
v -0.5 0.5 1.5
v -0.5 1.5 1.5
v -0.5 1.5 -0.5
v -0.5 0.5 -0.5
v -0.5 1.5 -0.5
v -0.5 1.5 1.5
v 0.5 1.5 1.5
v 0.5 1.5 -0.5
v -0.5 0.5 -0.5
v -0.5 1.5 -0.5
v 0.5 1.5 -0.5
v 0.5 0.5 -0.5
v 0.5 0.5 1.5
v 0.5 1.5 1.5
v -0.5 1.5 1.5
v -0.5 0.5 1.5
v -1.5 -1.5 -1.5
v -1.5 0.5 -1.5
v -0.5 0.5 -1.5
v -0.5 -1.5 -1.5
v 0.5 -0.5 -1.5
v 0.5 -1.5 -1.5
v -0.5 -1.5 -1.5
v -0.5 -0.5 -1.5
v -0.5 0.5 -1.5
v 0.5 0.5 -1.5
v 0.5 -0.5 -1.5
v -0.5 -0.5 -1.5
v -1.5 -1.5 0.5
v -0.5 -1.5 0.5
v -0.5 0.5 0.5
v -1.5 0.5 0.5
v -0.5 -1.5 0.5
v 0.5 -1.5 0.5
v 0.5 -0.5 0.5
v -0.5 -0.5 0.5
v -0.5 -0.5 0.5
v -0.5 -0.5 1.5
v -0.5 0.5 1.5
v -0.5 0.5 0.5
v 0.5 -0.5 0.5
v 0.5 -0.5 1.5
v -0.5 -0.5 1.5
v -0.5 -0.5 0.5
v -0.5 -0.5 1.5
v 0.5 -0.5 1.5
v 0.5 0.5 1.5
v -0.5 0.5 1.5
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
f 3//3 2//2 1//1
f 4//4 3//3 1//1
f 7//7 6//6 5//5
f 8//8 7//7 5//5
f 11//11 10//10 9//9
f 12//12 11//11 9//9
f 15//15 14//14 13//13
f 16//16 15//15 13//13
f 19//19 18//18 17//17
f 20//20 19//19 17//17
f 23//23 22//22 21//21
f 24//24 23//23 21//21
f 27//27 26//26 25//25
f 28//28 27//27 25//25
f 31//31 30//30 29//29
f 32//32 31//31 29//29
f 35//35 34//34 33//33
f 36//36 35//35 33//33
f 39//39 38//38 37//37
f 40//40 39//39 37//37
f 43//43 42//42 41//41
f 44//44 43//43 41//41
f 47//47 46//46 45//45
f 48//48 47//47 45//45
f 51//51 50//50 49//49
f 52//52 51//51 49//49
f 55//55 54//54 53//53
f 56//56 55//55 53//53
f 59//59 58//58 57//57
f 60//60 59//59 57//57
f 63//63 62//62 61//61
f 64//64 63//63 61//61
f 67//67 66//66 65//65
f 68//68 67//67 65//65
f 71//71 70//70 69//69
f 72//72 71//71 69//69
f 75//75 74//74 73//73
f 76//76 75//75 73//73
f 79//79 78//78 77//77
f 80//80 79//79 77//77
f 83//83 82//82 81//81
f 84//84 83//83 81//81
f 87//87 86//86 85//85
f 88//88 87//87 85//85
f 91//91 90//90 89//89
f 92//92 91//91 89//89
f 95//95 94//94 93//93
f 96//96 95//95 93//93
"
`;

exports[`csg invert, union, subtract, intersect 2`] = `
"v -1.5 -1.5 -1.5
v -1.5 -1.5 0.5
v -1.5 0.5 0.5
v -1.5 0.5 -1.5
v 0.5 -1.5 -1.5
v 0.5 -0.5 -1.5
v 0.5 -0.5 0.5
v 0.5 -1.5 0.5
v 0.5 -0.5 -0.5
v 0.5 -0.5 -1.5
v 0.5 0.5 -1.5
v 0.5 0.5 -0.5
v -1.5 -1.5 -1.5
v -0.5 -1.5 -1.5
v -0.5 -1.5 0.5
v -1.5 -1.5 0.5
v -0.5 -1.5 -1.5
v 0.5 -1.5 -1.5
v 0.5 -1.5 0.5
v -0.5 -1.5 0.5
v -0.5 0.5 -1.5
v -1.5 0.5 -1.5
v -1.5 0.5 0.5
v -0.5 0.5 0.5
v -0.5 0.5 -1.5
v -0.5 0.5 -0.5
v 0.5 0.5 -0.5
v 0.5 0.5 -1.5
v -0.5 -1.5 -1.5
v -1.5 -1.5 -1.5
v -1.5 0.5 -1.5
v -0.5 0.5 -1.5
v -0.5 -1.5 -1.5
v -0.5 -0.5 -1.5
v 0.5 -0.5 -1.5
v 0.5 -1.5 -1.5
v -0.5 -0.5 -1.5
v -0.5 0.5 -1.5
v 0.5 0.5 -1.5
v 0.5 -0.5 -1.5
v -1.5 -1.5 0.5
v -0.5 -1.5 0.5
v -0.5 0.5 0.5
v -1.5 0.5 0.5
v -0.5 -0.5 0.5
v -0.5 -1.5 0.5
v 0.5 -1.5 0.5
v 0.5 -0.5 0.5
v -0.5 0.5 -0.5
v -0.5 0.5 0.5
v -0.5 -0.5 0.5
v -0.5 -0.5 -0.5
v -0.5 -0.5 0.5
v 0.5 -0.5 0.5
v 0.5 -0.5 -0.5
v -0.5 -0.5 -0.5
v 0.5 -0.5 -0.5
v 0.5 0.5 -0.5
v -0.5 0.5 -0.5
v -0.5 -0.5 -0.5
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
f 3//3 2//2 1//1
f 4//4 3//3 1//1
f 7//7 6//6 5//5
f 8//8 7//7 5//5
f 11//11 10//10 9//9
f 12//12 11//11 9//9
f 15//15 14//14 13//13
f 16//16 15//15 13//13
f 19//19 18//18 17//17
f 20//20 19//19 17//17
f 23//23 22//22 21//21
f 24//24 23//23 21//21
f 27//27 26//26 25//25
f 28//28 27//27 25//25
f 31//31 30//30 29//29
f 32//32 31//31 29//29
f 35//35 34//34 33//33
f 36//36 35//35 33//33
f 39//39 38//38 37//37
f 40//40 39//39 37//37
f 43//43 42//42 41//41
f 44//44 43//43 41//41
f 47//47 46//46 45//45
f 48//48 47//47 45//45
f 51//51 50//50 49//49
f 52//52 51//51 49//49
f 55//55 54//54 53//53
f 56//56 55//55 53//53
f 59//59 58//58 57//57
f 60//60 59//59 57//57
"
`;

exports[`csg invert, union, subtract, intersect 3`] = `
"v 0.5 -0.5 -0.5
v 0.5 0.5 -0.5
v 0.5 0.5 0.5
v 0.5 -0.5 0.5
v -0.5 0.5 -0.5
v -0.5 0.5 0.5
v 0.5 0.5 0.5
v 0.5 0.5 -0.5
v -0.5 -0.5 0.5
v 0.5 -0.5 0.5
v 0.5 0.5 0.5
v -0.5 0.5 0.5
v -0.5 -0.5 -0.5
v -0.5 -0.5 0.5
v -0.5 0.5 0.5
v -0.5 0.5 -0.5
v -0.5 -0.5 -0.5
v 0.5 -0.5 -0.5
v 0.5 -0.5 0.5
v -0.5 -0.5 0.5
v -0.5 -0.5 -0.5
v -0.5 0.5 -0.5
v 0.5 0.5 -0.5
v 0.5 -0.5 -0.5
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
f 3//3 2//2 1//1
f 4//4 3//3 1//1
f 7//7 6//6 5//5
f 8//8 7//7 5//5
f 11//11 10//10 9//9
f 12//12 11//11 9//9
f 15//15 14//14 13//13
f 16//16 15//15 13//13
f 19//19 18//18 17//17
f 20//20 19//19 17//17
f 23//23 22//22 21//21
f 24//24 23//23 21//21
"
`;

exports[`csg invert, union, subtract, intersect 4`] = `
"v -1.5 0.5 -1.5
v -1.5 0.5 0.5
v -1.5 -1.5 0.5
v -1.5 -1.5 -1.5
v 0.5 -1.5 0.5
v 0.5 0.5 0.5
v 0.5 0.5 -1.5
v 0.5 -1.5 -1.5
v -1.5 -1.5 0.5
v 0.5 -1.5 0.5
v 0.5 -1.5 -1.5
v -1.5 -1.5 -1.5
v 0.5 0.5 -1.5
v 0.5 0.5 0.5
v -1.5 0.5 0.5
v -1.5 0.5 -1.5
v 0.5 -1.5 -1.5
v 0.5 0.5 -1.5
v -1.5 0.5 -1.5
v -1.5 -1.5 -1.5
v -1.5 0.5 0.5
v 0.5 0.5 0.5
v 0.5 -1.5 0.5
v -1.5 -1.5 0.5
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn 1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn -1 0 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 -1 0
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
vn 0 0 -1
f 3//3 2//2 1//1
f 4//4 3//3 1//1
f 7//7 6//6 5//5
f 8//8 7//7 5//5
f 11//11 10//10 9//9
f 12//12 11//11 9//9
f 15//15 14//14 13//13
f 16//16 15//15 13//13
f 19//19 18//18 17//17
f 20//20 19//19 17//17
f 23//23 22//22 21//21
f 24//24 23//23 21//21
"
`;
